ksp_plugin/part.cpp

```diff
@@ -161,56 +161,50 @@ RigidMotion<RigidPart, Barycentric> const& Part::rigid_motion() const {
   return rigid_motion_;
 }
 
-DiscreteTrajectory<Barycentric>::Iterator Part::history_begin() {
-  // Make sure that we skip the point of the prehistory.
-  auto it = history_->Fork();
-  return ++it;
+DiscreteTraject0ry<Barycentric>::iterator Part::history_begin() {
+  return history_->begin();
 }
 
-DiscreteTrajectory<Barycentric>::Iterator Part::history_end() {
+DiscreteTraject0ry<Barycentric>::iterator Part::history_end() {
   return history_->end();
 }
 
-DiscreteTrajectory<Barycentric>::Iterator Part::psychohistory_begin() {
-  if (psychohistory_ == nullptr) {
-    psychohistory_ = history_->NewForkAtLast();
+DiscreteTraject0ry<Barycentric>::iterator Part::psychohistory_begin() {
+  if (psychohistory_ == trajectory_.segments().end()) {
+    psychohistory_ = trajectory_.NewSegment();
   }
+  // TODO(phl): This used to say:
   // Make sure that we skip the fork, which may be the point of the prehistory.
-  auto it = psychohistory_->Fork();
-  return ++it;
+  return psychohistory_->begin();
 }
 
-DiscreteTrajectory<Barycentric>::Iterator Part::psychohistory_end() {
-  if (psychohistory_ == nullptr) {
-    psychohistory_ = history_->NewForkAtLast();
+DiscreteTraject0ry<Barycentric>::iterator Part::psychohistory_end() {
+  if (psychohistory_ == trajectory_.segments().end()) {
+    psychohistory_ = trajectory_.NewSegment();
   }
   return psychohistory_->end();
 }
 
 void Part::AppendToHistory(
     Instant const& time,
     DegreesOfFreedom<Barycentric> const& degrees_of_freedom) {
-  if (psychohistory_ != nullptr) {
-    history_->DeleteFork(psychohistory_);
+  if (psychohistory_ != trajectory_.segments().end()) {
+    trajectory_.DeleteSegments(psychohistory_);
   }
-  history_->Append(time, degrees_of_freedom);
+  trajectory_.Append(time, degrees_of_freedom);
 }
 
 void Part::AppendToPsychohistory(
     Instant const& time,
     DegreesOfFreedom<Barycentric> const& degrees_of_freedom) {
-  if (psychohistory_ == nullptr) {
-    psychohistory_ = history_->NewForkAtLast();
+  if (psychohistory_ == trajectory_.segments().end()) {
+    psychohistory_ = trajectory_.NewSegment();
   }
-  psychohistory_->Append(time, degrees_of_freedom);
+  trajectory_.Append(time, degrees_of_freedom);
 }
 
 void Part::ClearHistory() {
-  if (psychohistory_ != nullptr) {
-    history_->DeleteFork(psychohistory_);
-  }
-  prehistory_->DeleteFork(history_);
-  history_ = prehistory_->NewForkAtLast();
+  trajectory_ = DiscreteTraject0ry<Barycentric>();
```

Comment:
Shouldn’t this set `psychohistory_` to `trajectory_.segments().end()`?

Reply:
Good catch, done.  This should also call `clear`, not do an assignment, to preserve `history_`.